README.md
# Redhat Communties of Practice API Framework Collection

![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)
![Code style: flake8](https://img.shields.io/badge/Code%20style-flake8-orange)
<!-- Further CI badges go here as above -->

This is a Framework Collection only. The purpose is to provide 80% of the code needed to interact with any given api. It is based off the Ansible Tower/Automation Hub Api, but most of the functions should be relatively close to what would be needed. This provides a framework for creating modules that will interact with an API.

This Ansible collection provides a framework for a collection to allow for easy interaction with an API via Ansible playbooks.

This code will not work in its current form. Many of the functions will need tweaked in order to work correctly. For example assocations may not be present in a particular API, or they may use other naming/id conventions.

The rest of this Readme also provides a framework for content that should be in a collection Readme, and is meant as an example only.

## Included content

Click the `Content` button to see the list of content included in this collection.

## Installing this collection

You can install the redhat_cop api_framework collection with the Ansible Galaxy CLI:

    ansible-galaxy collection install redhat_cop.api_framework

You can also include it in a `requirements.yml` file and install it with `ansible-galaxy collection install -r requirements.yml`, using the format:

```yaml
---
collections:
  - name: redhat_cop.api_framework
    # If you need a specific version of the collection, you can specify like this:
    # version: ...
```

## Using this collection
Define following vars here, or in `configs/auth.yml`
`server: http://apiserver.com`

You can also specify authentication by setting the following variables:
 - `token`, `oauthtoken`

The OAuth2 token is the only method. You can obtain the token through through the web interface if applicable

These can be specified via (from highest to lowest precedence):

 - direct role variables as mentioned above

### See Also:

* [Ansible Using collections](https://docs.ansible.com/ansible/latest/user_guide/collections_using.html) for more details.

## Release and Upgrade Notes
For details on changes between versions, please see [the changelog for this collection](CHANGELOG.rst).

## Contributing to this collection

We welcome community contributions to this collection. If you find problems, please open an issue or create a PR against the [API Framework collection repository](https://github.com/redhat-cop/api_framework).
More information about contributing can be found in our [Contribution Guidelines.](https://github.com/redhat-cop/api_framework/blob/devel/.github/CONTRIBUTING.md)

## Licensing
GNU General Public License v3.0 or later.

See [LICENCE](https://www.gnu.org/licenses/gpl-3.0.txt) to see the full text.

All content in this folder is licensed under the same license as Ansible,
which is the same as license that applied before when the base for this
code was derived form the [AWX.AWX](https://galaxy.ansible.com/awx/awx) collection.


CHANGELOG.rst
=========================================
redhat_cop.ah_configuration Release Notes
=========================================

.. contents:: Topics


v0.1.0
======

Major Changes
-------------

- Initial release of api framework


plugins/modules/token.py
#!/usr/bin/python
# coding: utf-8 -*-


# (c) 2020, John Westcott IV <john.westcott.iv@redhat.com>
# (c) 2021, Sean Sullivan <@sean-m-sullivan>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

__metaclass__ = type


ANSIBLE_METADATA = {"metadata_version": "1.1", "status": ["preview"], "supported_by": "community"}

DOCUMENTATION = """
---
module: token
author: "John Westcott IV (@john-westcott-iv), Sean Sullivan (@sean-m-sullivan)"
version_added: "0.0.0"
short_description: create, update, or destroy tokens.
description:
    - Create or destroy tokens.
    - In addition, the module sets an Ansible fact which can be passed into other
      modules as the parameter oauthtoken. See examples for usage.
    - Because of the sensitive nature of tokens, the created token value is only available once
      through the Ansible fact. (See RETURN for details)
    - Due to the nature of tokens this module is not idempotent. A second will
      with the same parameters will create a new token.
    - If you are creating a temporary token for use with modules you should delete the token
      when you are done with it. See the example for how to do it.
options:
    state:
      description:
        - Desired state of the resource.
      choices: ["present", "absent"]
      default: "present"
      type: str
extends_documentation_fragment: redhat_cop.api_framework.auth
"""

EXAMPLES = """
- name: Create a new token using an existing token
  token:
    token: "{{ my_existing_token }}"

- name: Delete this token
  token:
    token: "{{ token }}"
    state: absent

- name: Create a new token using username/password
  token:
    state: present
    username: "{{ my_username }}"
    password: "{{ my_password }}"

- name: Use our new token to make another call
  namespace:
    token: "{{ token }}"
"""

RETURN = """
token:
  type: dict
  description: An Ansible Fact variable representing a Tower token object which can be used for auth in subsequent modules. See examples for usage.
  contains:
    token:
      description: The token that was generated. This token can never be accessed again, make sure this value is noted before it is lost.
      type: str
    id:
      description: The numeric ID of the token created
      type: str
  returned: on successful create
"""

from ..module_utils.module import Module


def return_token(module, last_response):
    # A token is special because you can never get the actual token ID back from the API.
    # So the default module return would give you an ID but then the token would forever be masked on you.
    # This method will return the entire token object we got back so that a user has access to the token

    module.json_output["ansible_facts"] = {
        # 'token': last_response['token'],
        "token": last_response,
    }
    module.exit_json(**module.json_output)


def main():
    # Any additional arguments that are not fields of the item can be added here
    argument_spec = dict(
        state=dict(choices=["present", "absent"], default="present"),
    )

    # Create a module for ourselves
    module = Module(argument_spec=argument_spec)

    # Extract our parameters
    state = module.params.get("state")

    # Delete an existing token
    if state == "absent":
        # If the state was absent we can let the module delete it if needed, the module will handle exiting from this
        existing_item = {}
        existing_item["endpoint"] = "auth/token/"
        existing_item["type"] = "token"
        module.delete_if_needed(existing_item)

    # If the state was present and we can let the module build or update the existing item, this will return on its own
    module.create_or_update_if_needed(
        None,
        None,
        endpoint="auth/token/",
        item_type="token",
        on_create=return_token,
    )


if __name__ == "__main__":
    main()


plugins/doc_fragments/auth.py
# -*- coding: utf-8 -*-

# Copyright: (c) 2020, Sean Sullivan <@sean-m-sullivan>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

__metaclass__ = type


class ModuleDocFragment(object):

    # Ansible Galaxy documentation fragment
    DOCUMENTATION = r"""
options:
  host:
    description:
    - URL to Host instance.
    - If value not set, will try environment variable C(HOST)
    - If value not specified by any means, the value of C(127.0.0.1) will be used
    type: str
  username:
    description:
    - Username for your Host instance.
    - If value not set, will try environment variable C(USERNAME)
    type: str
  password:
    description:
    - Password for your Host instance.
    - If value not set, will try environment variable C(PASSWORD)
    type: str
  token:
    description:
    - The Host API token to use.
    - This value can be in one of two formats.
    - A string which is the token itself. (i.e. bqV5txm97wqJqtkxlMkhQz0pKhRMMX)
    - A dictionary structure as returned by the token module.
    - If value not set, will try environment variable C(API_TOKEN)
    type: raw
  validate_certs:
    description:
    - Whether to allow insecure connections to Galaxy or Automation Hub Server.
    - If C(no), SSL certificates will not be validated.
    - This should only be used on personally controlled sites using self-signed certificates.
    - If value not set, will try environment variable C(VERIFY_SSL)
    type: bool
    aliases: [ verify_ssl ]
"""


plugins/modules/namespace.py
#!/usr/bin/python
# coding: utf-8 -*-

# (c) 2020, Sean Sullivan <@sean-m-sullivan>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

__metaclass__ = type


ANSIBLE_METADATA = {"metadata_version": "1.1", "status": ["preview"], "supported_by": "community"}


DOCUMENTATION = """
---
module: namespace
author: "Sean Sullivan (@sean-m-sullivan)"
short_description: create, update, or destroy Automation Hub Namespace.
description:
    - Create, update, or destroy Automation Hub Namespace. See
      U(https://www.ansible.com/) for an overview.
options:
    name:
      description:
        - Namespace name. Must be lower case containing only alphanumeric characters and underscores.
      required: True
      type: str
    new_name:
      description:
        - Setting this option will change the existing name (looked up via the name field.
      type: str
    description:
      description:
        - Description to use for the Namespace.
      type: str
    company:
      description:
        - Namespace owner company name.
      type: str
    email:
      description:
        - Namespace contact email.
      type: str
    avatar_url:
      description:
        - Namespace logo URL.
      type: str
    resources:
      description:
        - Namespace resource page in Markdown format.
      type: str
    state:
      description:
        - Desired state of the resource.
        - Currently the ability to delete objects in Automation Hub is not available, this option is included for when it is.
      choices: ["present"]
      default: "present"
      type: str
    links:
      description:
        - A list of dictionaries of Name and url values for links related the Namespace.
      type: list
      elements: dict
      suboptions:
        name:
          description:
            - Link Text.
          type: str
          required: True
        url:
          description:
            - Link URL.
          type: str
          required: True
    groups:
      description:
        - A list of dictionaries of the Names and object_permissions values for groups that control the Namespace.
      type: list
      required: True
      elements: dict
      suboptions:
        name:
          description:
            - Group Name or ID.
          type: str
          required: True
        object_permissions:
          description:
            - List of Permisions granted to the group.
          choices: ["change_namespace", "upload_to_namespace"]
          type: list
          required: True

extends_documentation_fragment: redhat_cop.api_framework.auth
"""


EXAMPLES = """
- name: Create Tower Ping job template
  namespace:
    name: Redhat
    company: Redhat
    email: user@example.com
    avatar_url: https://pnt.redhat.com/pnt/d-11633955/LogoRedHatHatColorRGB.png
    description: This is the Redhat Namespace
    links:
      - name: "homepage"
        url: "http://www.redhat.com"
    groups:
      - name: system:partner-engineers
        object_permissions:
          - "change_namespace"
          - "upload_to_namespace"

"""

from ..module_utils.module import Module


def main():
    # Any additional arguments that are not fields of the item can be added here
    argument_spec = dict(
        name=dict(required=True),
        new_name=dict(),
        description=dict(),
        company=dict(),
        email=dict(),
        avatar_url=dict(),
        resources=dict(),
        links=dict(type="list", elements="dict"),
        groups=dict(required=True, type="list", elements="dict"),
        state=dict(choices=["present"], default="present"),
    )

    # Create a module for ourselves
    module = Module(argument_spec=argument_spec)

    # Extract our parameters
    name = module.params.get("name")
    new_name = module.params.get("new_name")
    state = module.params.get("state")

    new_fields = {}

    # Attempt to look up an existing item based on the provided data
    existing_item = module.get_one("namespaces", name_or_id=name)

    if state == "absent":
        # If the state was absent we can let the module delete it if needed, the module will handle exiting from this
        module.delete_if_needed(existing_item)

    # Create the data that gets sent for create and update
    # Remove these two comments for final
    # Check that Links and groups works with this.
    new_fields["name"] = new_name if new_name else (module.get_item_name(existing_item) if existing_item else name)
    for field_name in (
        "description",
        "company",
        "email",
        "avatar_url",
        "resources",
        "links",
        "groups",
    ):
        field_val = module.params.get(field_name)
        if field_val is not None:
            new_fields[field_name] = field_val

    # If the state was present and we can let the module build or update the existing item, this will return on its own
    module.create_or_update_if_needed(
        existing_item,
        new_fields,
        endpoint="namespaces",
        item_type="namespaces",
    )


if __name__ == "__main__":
    main()
